Add Copilot feature

Guardiola31337 opened the pull request:
### Description
<!--
Include issue references (e.g., fixes [#issue](link))
Include necessary implementation details (e.g. I opted to use this algorithm because ... and test it in this way ...).
-->
- Adds Copilot feature

~~Marking this as a `Draft` and adding `skip changelog` label while we work on it.~~

cc @AhmerKhan1 @gmoedl @zugaldia @brandanblevins

Guardiola31337 commented:
Unrelated to this PR but `additional_alternative_is_not_force_to_invoke_alternatives_observer` failed https://console.firebase.google.com/u/0/project/mapbox-navigation-android/testlab/histories/bh.54eb2dd57e959221/matrices/8304525423919644726/executions/bs.2430a376891536d5/testcases/27

```
java.lang.RuntimeException: java.util.concurrent.ExecutionException: androidx.test.espresso.IdlingResourceTimeoutException: Wait for [RouteAlternativesIdlingResource] to become idle timed out
	at androidx.test.espresso.Espresso.onIdle(Espresso.java:15)
	at androidx.test.espresso.Espresso.onIdle(Espresso.java:1)
	at com.mapbox.navigation.instrumentation_tests.core.RouteAlternativesTest$additional_alternative_is_not_force_to_invoke_alternatives_observer$3.invoke(RouteAlternativesTest.kt:239)
	at com.mapbox.navigation.instrumentation_tests.core.RouteAlternativesTest$additional_alternative_is_not_force_to_invoke_alternatives_observer$3.invoke(RouteAlternativesTest.kt:238)
	at com.mapbox.navigation.instrumentation_tests.utils.history.MapboxHistoryTestRule.stopRecordingOnCrash(MapboxHistoryTestRule.kt:71)
	at com.mapbox.navigation.instrumentation_tests.core.RouteAlternativesTest.additional_alternative_is_not_force_to_invoke_alternatives_observer(RouteAlternativesTest.kt:238)
	... 42 trimmed
Caused by: java.util.concurrent.ExecutionException: androidx.test.espresso.IdlingResourceTimeoutException: Wait for [RouteAlternativesIdlingResource] to become idle timed out
	at java.util.concurrent.FutureTask.report(FutureTask.java:94)
	at java.util.concurrent.FutureTask.get(FutureTask.java:164)
	at androidx.test.espresso.Espresso.onIdle(Espresso.java:10)
	... 48 more
Caused by: androidx.test.espresso.IdlingResourceTimeoutException: Wait for [RouteAlternativesIdlingResource] to become idle timed out
	at androidx.test.espresso.IdlingPolicy.handleTimeout(IdlingPolicy.java:4)
	at androidx.test.espresso.base.UiControllerImpl$5.resourcesHaveTimedOut(UiControllerImpl.java:1)
	at androidx.test.espresso.base.IdlingResourceRegistry$Dispatcher.handleTimeout(IdlingResourceRegistry.java:4)
	at androidx.test.espresso.base.IdlingResourceRegistry$Dispatcher.handleMessage(IdlingResourceRegistry.java:6)
	at android.os.Handler.dispatchMessage(Handler.java:98)
	at androidx.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:14)
	at androidx.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:8)
	at androidx.test.espresso.base.UiControllerImpl.loopMainThreadUntilIdle(UiControllerImpl.java:17)
	at androidx.test.espresso.Espresso$1.run(Espresso.java:1)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:423)
	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
	at android.os.Handler.handleCallback(Handler.java:739)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:148)
	at android.app.ActivityThread.main(ActivityThread.java:5417)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```

cc @dzinad @VysotskiVadim @RingerJK

Guardiola31337 commented:
@mapbox/navigation-android take a look at the PR and share your thoughts when you have a chance 🙏

In parallel, I'm going to work on documentation and examples but "core" logic is ready for review.

Zayankovsky commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/copilot/HistoryAttachmentsUtils.kt:
```diff
@@ -0,0 +1,86 @@
+        } else {
+            JSONObject(
+                String(
+                    Base64.decode(
```
We can use `android.util.Base64` in all API levels, it is not deprecated.

Zayankovsky replied:
In tests you can do something like this:
```kotlin
mockkStatic(Base64::class)
every { Base64.decode(any<String>(), any()) } answers {
    getDecoder().decode(firstArg<String>())
}
```

Guardiola31337 replied:
Let me try that out 👍

Guardiola31337 replied:
It works 🎉 

TIL

Zayankovsky commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:
```diff
@@ -1097,6 +1099,29 @@ class MapboxNavigation @VisibleForTesting internal constructor(
+    /**
+     * start
+     */
+    fun startCopilot() {
+        MapboxCopilot.start()
+    }
```
Does this have to be in `MapboxNavigation`? Why not expose `MapboxCopilot` instead, so that it can be started independently?

kmadsen replied:
Maybe it's because MapboxCopilot does not have a public interface. MapboxNavigation creates a public reference point. It also makes it so you must set up NavigationOptions before you can use copilot.

Guardiola31337 replied:
Correct. Also, IMO having these under `MapboxNavigation` makes the feature more discoverable.

Zayankovsky replied:
I think having this in `MapboxNavigation` creates a false assumption that Copilot will be automatically stopped when this `MapboxNavigation` instance is destroyed, but in fact it will automatically resume when a new instance is created, because Copilot uses `MapboxNavigationObserver` under the hood.

Guardiola31337 replied:
This is a great point 🤔 

@kmadsen and @mapbox/navigation-android any other opinions on the discussion of having Copilot within `MapboxNavigation` or not?

dzinad replied:
I like the idea of moving Copilot API to a separate class.
If you want to keep `MapboxNavigation` one single entry point, you can at least introduce `MapboxNavigation#getCopilot` method that would return an object with all the Copilot API.
But actually `MapboxNavigation` is not a single entry point now. Look at any example: there are `MapboxRouteLineView`, `MapboxRouteLineApi`, `MapboxVoiceInstructionsPlayer` and other things. Why can't Copoilot be one of them?

Zayankovsky reviewed (commented):
To be honest, I don't see any benefits of moving Copilot into Nav SDK, if users still have to manually start and stop history recording.

kmadsen commented on libnavigation-core/api/current.txt:
```diff
@@ -173,6 +176,117 @@ package com.mapbox.navigation.core.arrival {
+    property public final com.mapbox.navigation.core.copilot.SearchResultUsed searchResultUsed;
+  }
+
+  @Keep public final class SearchResults {
```
Do you want to start with @Experimental annotations while it is being integrated further?

Just in case there are some new learnings from wider adoption, you'll at least be able to make minor tweaks

Guardiola31337 replied:
Yeah, good point. Was planning to mark any Copilot API as `@Experimental` until we go GA 👍 

Opened this as a `Draft` to get some initial feedback from you but yeah, APIs are going to be `@Experimental` for now.

Guardiola31337 replied:
Added `@ExperimentalPreviewMapboxNavigationAPI` annotations to all public entities 👍

kmadsen commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:
```diff
@@ -1097,6 +1099,29 @@ class MapboxNavigation @VisibleForTesting internal constructor(
+    /**
+     * start
+     */
+    fun startCopilot() {
```
```suggestion
    fun startCopilot() = apply {
```

IMO it is nice to be able to chain functions

Guardiola31337 commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:
```diff
@@ -1097,6 +1099,29 @@ class MapboxNavigation @VisibleForTesting internal constructor(
     }
 
+    /**
+     * start
```
Sure thing!

> In parallel, I'm going to work on documentation and examples but "core" logic is ready for review.

Opened this as a `Draft` to get some initial feedback from you but yeah, currently working on documentation and examples.

Guardiola31337 commented:
@Zayankovsky 
> To be honest, I don't see any benefits of moving Copilot into Nav SDK, if users still have to manually start and stop history recording.

The idea was to have Copilot bundled as a feature of the Nav SDK, so that users don't need to do anything from their side, removing any friction which will greatly help with adoption.

Zayankovsky commented:
> The idea was to have Copilot bundled as a feature of the Nav SDK, so that users don't need to do anything from their side, removing any friction which will greatly help with adoption.

They still have to call startCopilot and stopCopilot, which means the only profit for users is that they don’t have to add Copilot dependency, which is just a single line in a cradle script.

Guardiola31337 commented:
@Zayankovsky @kmadsen @dzinad addressed the feedback based on above comments and internal discussions. This is ready for another round of 👀 

Only missing thing is https://github.com/mapbox/mapbox-navigation-android/pull/6572#discussion_r1018175392 which I'm currently working on 🚀

kmadsen commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilot.kt:
```diff
@@ -0,0 +1,66 @@
+/**
+ * MapboxCopilot.
+ */
+object MapboxCopilot : MapboxNavigationObserver {
```
```suggestion
@ExperimentalPreviewMapboxNavigationAPI
object MapboxCopilot : MapboxNavigationObserver {
```

Was there a reason to call the individual functions experimental rather than the top level?

Zayankovsky replied:
I think it is better to make the whole class experimental, so that in the future we will be able to move it to a separate package, rename or completely remove if necessary. With the current implementation we will be able to change only its functions.

Guardiola31337 replied:
Great point. Let me quickly fix and re-push.

kmadsen commented on libnavigation-copilot/src/test/resources/robolectric.properties:
```diff
@@ -0,0 +1 @@
+sdk=28
```
TIL

kmadsen commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/SearchResults.kt:
```diff
@@ -0,0 +1,25 @@
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+
+/**
+ * SearchResults
```
Will you add description here?

kmadsen commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/SearchResultUsed.kt:
```diff
@@ -0,0 +1,25 @@
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+
+/**
+ * SearchResultUsed
+ *
+ * @property provider
+ * @property id
+ * @property name
+ * @property address
+ * @property coordinates
+ * @property routablePoint
```
Same here, can there be more description here? Maybe, what does "Used" mean in this context. Is it, user selected or what?

Guardiola31337 commented:
Unrelated to this PR but `history_recording_observer_ensures_first_set_route_event` failed https://console.firebase.google.com/u/0/project/mapbox-navigation-android/testlab/histories/bh.54eb2dd57e959221/matrices/6503567225429387593/executions/bs.288fb8cabf5f81b7

```
java.lang.NullPointerException
at com.mapbox.navigation.instrumentation_tests.core.HistoryRecordingStateChangeObserverTest$history_recording_observer_ensures_first_set_route_event$1.invokeSuspend(HistoryRecordingStateChangeObserverTest.kt:354)
at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
at kotlinx.coroutines.DispatchedTaskKt.resume(DispatchedTask.kt:234)
at kotlinx.coroutines.DispatchedTaskKt.resumeUnconfined(DispatchedTask.kt:190)
at kotlinx.coroutines.DispatchedTaskKt.dispatch(DispatchedTask.kt:161)
at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume(CancellableContinuationImpl.kt:397)
at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl(CancellableContinuationImpl.kt:431)
at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl$default(CancellableContinuationImpl.kt:420)
at kotlinx.coroutines.CancellableContinuationImpl.resumeWith(CancellableContinuationImpl.kt:328)
at com.mapbox.navigation.instrumentation_tests.core.HistoryRecordingStateChangeObserverTest$awaitStopActiveGuidanceRecording$2$observer$1$onShouldStopRecording$1.onSaved(HistoryRecordingStateChangeObserverTest.kt:462)
at com.mapbox.navigation.core.history.MapboxHistoryRecorder.stopRecording$lambda-2$lambda-1(MapboxHistoryRecorder.kt:63)
at com.mapbox.navigation.core.history.MapboxHistoryRecorder.$r8$lambda$YnMgKxAxhgNPiSCf6sQIkMwHe9U(Unknown Source:0)
at com.mapbox.navigation.core.history.MapboxHistoryRecorder$$ExternalSyntheticLambda0.run(Unknown Source:2)
at android.os.MessageQueue.nativePollOnce(Native Method)
at android.os.MessageQueue.next(MessageQueue.java:335)
at android.os.Looper.loopOnce(Looper.java:186)
at android.os.Looper.loop(Looper.java:313)
at android.app.ActivityThread.main(ActivityThread.java:8663)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:567)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1135)
```

cc @dzinad as it seems a regression from https://github.com/mapbox/mapbox-navigation-android/pull/6525

dzinad commented:
> cc @dzinad as it seems a regression from https://github.com/mapbox/mapbox-navigation-android/pull/6525

No, that's exactly what this PR was supposed to fix. Probably it didn't.

Guardiola31337 commented:
@dzinad 
> UPD: did you rebase? I don't see the fix in your branch.

I'd think so but let me pull the latest changes from `main` and re-run. BTW what about https://github.com/mapbox/mapbox-navigation-android/pull/6572#issuecomment-1307815992?

Guardiola31337 commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/HistoryEvents.kt:
```diff
@@ -0,0 +1,64 @@
+ * @see [HistoryEvent]
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+data class SearchResultUsedEvent(val searchResultUsed: SearchResultUsed) :
```
This was to avoid the need of opt-in `ExperimentalPreviewMapboxNavigationAPI` in some other internal classes (e.g. any class that relies on `MapboxCopilot.pushStatusObservers`.

Guardiola31337 replied:
2 main reasons:
- We cannot generate the events internally as we don't integrate the Search SDK as part of the Nav SDK (yet)
- We also have that open for other search providers

dzinad commented:
> I'd think so but let me pull the latest changes from main and re-run. BTW what about https://github.com/mapbox/mapbox-navigation-android/pull/6572#issuecomment-1307815992?

Discussing in NAVAND-583.

Guardiola31337 commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/InitRoute.kt:
```diff
@@ -0,0 +1,10 @@
+import androidx.annotation.Keep
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+
+@Keep
```
If not events are minified in `release` builds and not readable afterwards.

Guardiola31337 replied:
When analyzing the traces I mean.

Guardiola31337 commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/SearchResults.kt:
```diff
@@ -0,0 +1,25 @@
+ */
+@Keep
+@ExperimentalPreviewMapboxNavigationAPI
+data class SearchResults(
```
https://github.com/mapbox/mapbox-navigation-android/pull/6572#discussion_r1019321968
Added `@ExperimentalPreviewMapboxNavigationAPI` annotation to all public entities. This shouldn't be a blocker for now.

Guardiola31337 commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilotImpl.kt:
```diff
@@ -0,0 +1,392 @@
+        }
+    private val historyRecordingStateChangeObserver = object : HistoryRecordingStateChangeObserver {
+        override fun onShouldStartRecording(state: HistoryRecordingSessionState) {
+            if (state !is Idle) {
```
Yeah, I know but was wondering if marking it explicitly makes it more understandable from a client standpoint. It also prevents failures in the future if that logic ever changes (although that'd be a behavioral breaking change). What do you think? 🤔

dzinad replied:
Yeah, sure, it wouldn't hurt to check twice.

Guardiola31337 commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/UploadServiceInterfaceFactory.kt:
```diff
@@ -0,0 +1,10 @@
+
+internal object UploadServiceInterfaceFactory {
+
+    fun retrieveUploadServiceInterface(): UploadServiceInterface =
```
Nah, you get `UnsatisfiedLinkError`s if not wrapped 💥 

```
'com.mapbox.common.UploadServiceInterface com.mapbox.common.UploadServiceFactory.getInstance()'
java.lang.UnsatisfiedLinkError: 'com.mapbox.common.UploadServiceInterface com.mapbox.common.UploadServiceFactory.getInstance()'
	at com.mapbox.common.UploadServiceFactory.getInstance(Unknown Source)
	at com.mapbox.navigation.copilot.HistoryUploadWorkerTest$prepareUploadMockks$4.invoke(HistoryUploadWorkerTest.kt:420)
	at com.mapbox.navigation.copilot.HistoryUploadWorkerTest$prepareUploadMockks$4.invoke(HistoryUploadWorkerTest.kt:419)
	at io.mockk.impl.eval.RecordedBlockEvaluator$record$block$1.invoke(RecordedBlockEvaluator.kt:25)
	at io.mockk.impl.eval.RecordedBlockEvaluator$enhanceWithRethrow$1.invoke(RecordedBlockEvaluator.kt:78)
	at io.mockk.impl.recording.JvmAutoHinter.autoHint(JvmAutoHinter.kt:23)
	at io.mockk.impl.eval.RecordedBlockEvaluator.record(RecordedBlockEvaluator.kt:40)
	at io.mockk.impl.eval.EveryBlockEvaluator.every(EveryBlockEvaluator.kt:30)
	at io.mockk.MockKDsl.internalEvery(API.kt:92)
	at io.mockk.MockKKt.every(MockK.kt:98)
	at com.mapbox.navigation.copilot.HistoryUploadWorkerTest.prepareUploadMockks(HistoryUploadWorkerTest.kt:419)
	at com.mapbox.navigation.copilot.HistoryUploadWorkerTest.access$prepareUploadMockks(HistoryUploadWorkerTest.kt:35)
	at com.mapbox.navigation.copilot.HistoryUploadWorkerTest$Result success - FINISHED 204$1.invokeSuspend(HistoryUploadWorkerTest.kt:256)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at com.mapbox.navigation.copilot.HistoryUploadWorkerTest.Result success - FINISHED 204(HistoryUploadWorkerTest.kt:247)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy5.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$2.run(TestWorker.java:176)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```

Zayankovsky commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/HistoryAttachmentsUtils.kt:
```diff
@@ -0,0 +1,86 @@
+        return to
+    }
+
+    @SuppressLint("NewApi")
```
Suppress is necessary here, because API level check is extracted to a separate function and lint can not understand that.

Guardiola31337 replied:
I'm not getting any warnings after implementing what @dzinad suggested 🤔

Zayankovsky replied:
Using `TargetApi` is basically the same as suppressing, you won't have any warnings now, even if you remove API level check. Anyway, it is better to simply use `android.util.Base64` with all API levels.

Zayankovsky reviewed (commented):
Looks good to me, will approve when the PR is complete.

Guardiola31337 commented:
@dzinad @kmadsen @Zayankovsky I've just pushed a commit addressing all the comments. This is ready for review 🚀

Zayankovsky commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilot.kt:
```diff
@@ -0,0 +1,95 @@
+     * @param mapboxNavigation instance that is being attached
+     */
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        MapboxCopilotImpl(mapboxNavigation).also {
+            copilot = it
+            it.start()
+        }
```
```suggestion
        copilot = MapboxCopilotImpl(mapboxNavigation).also { it.start() }
```

Zayankovsky commented on libnavigation-copilot/src/test/java/com/mapbox/navigation/copilot/HistoryAttachmentsUtilsTest.kt:
```diff
@@ -0,0 +1,74 @@
+
+    @Test
+    fun decode() {
+        mockkStatic(Base64::class)
+        every { Base64.decode(any<String>(), any()) } answers {
+            getDecoder().decode(firstArg<String>())
+        }
+        val accessToken =
+            "pk.eyJ1IjoiY29waWxvdC10ZXN0LW93bmVyIiwiYSI6ImNpdDF3OHpoaTAwMDcyeXA5Y3Z0Nmk2dzEifQ." +
+                "cit1w8zhi00072yp9cvt6i6w1"
+
+        val owner = HistoryAttachmentsUtils.retrieveOwnerFrom(accessToken)
+
+        assertEquals("copilot-test-owner", owner)
+        unmockkStatic(Base64::class)
+    }
```
```suggestion
        mockkStatic(Base64::class) {
            every { Base64.decode(any<String>(), any()) } answers {
                getDecoder().decode(firstArg<String>())
            }
            val accessToken =
                "pk.eyJ1IjoiY29waWxvdC10ZXN0LW93bmVyIiwiYSI6ImNpdDF3OHpoaTAwMDcyeXA5Y3Z0Nmk2dzEifQ." +
                    "cit1w8zhi00072yp9cvt6i6w1"

            val owner = HistoryAttachmentsUtils.retrieveOwnerFrom(accessToken)

            assertEquals("copilot-test-owner", owner)
        }
    }
```
This way the mock will be cleared even when test fails.

Zayankovsky commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilotImpl.kt:
```diff
@@ -0,0 +1,392 @@
+        val from = File(filePath)
+        val to = copyToAndRemove(from, metadata.name)
+        var url = if (isAppDebuggable()) {
+            STAGING_BASE_URL
```
Is it correct for external apps to send data to a staging host?

Guardiola31337 replied:
This is a really great question 🤔 

Could we restrict to Nav SDK debuggable?

Guardiola31337 replied:
Maybe we can rely on `BuildConfig.DEBUG` from the SDK (`libnavigation-core`) that only applies to us (as published artifacts are `release`), and then, the rest will be sent to the `production` endpoint. What do you think?

Zayankovsky replied:
It think it is fine to use `BuildConfig.DEBUG`. But we should probably discuss this with the team that owns these hosts. Also I don't know how useful traces from debuggable apps usually are.

Guardiola31337 replied:
This is pretty useful for testing purposes. IMO we should keep it.

Cc @DorkMatter

Guardiola31337 commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/HistoryPoint.kt:
```diff
@@ -0,0 +1,17 @@
+ */
+@Keep
+@ExperimentalPreviewMapboxNavigationAPI
+data class HistoryPoint(
```
Nope. I considered that when we started working on this. Problem is that `Point` has more fields that we don't want to include as part of the serialized events.

Zayankovsky commented on libnavigation-base/src/main/java/com/mapbox/navigation/base/extensions/NavigationOptionsEx.kt:
```diff
@@ -0,0 +1,20 @@
+import com.mapbox.navigation.base.options.CopilotOptions
+import com.mapbox.navigation.base.options.NavigationOptions
+
+/**
+ * Defines configuration for Copilot.
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+fun NavigationOptions.Builder.copilotOptions(copilotOptions: CopilotOptions):
+    NavigationOptions.Builder = this.copilotOptions(copilotOptions)
+
+/**
+ * Copilot configuration.
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+fun NavigationOptions.copilotOptions(): CopilotOptions = this.copilotOptions
```
Looks like we don't need these extensions, we can make corresponding functions and properties in `NavigationOptions` public and add annotations there.

Guardiola31337 replied:
Thought about that and was wondering if that approach would have more impact from the client side if we have to change the API in the future, causing more affecting changes to developers. Let me take it for a quick spin and double check 👍

Guardiola31337 replied:
Fixed and pushed. Ready for a final round of 👀

Zayankovsky commented on libnavigation-base/src/main/java/com/mapbox/navigation/base/options/NavigationOptions.kt:
```diff
@@ -62,7 +63,8 @@ class NavigationOptions private constructor(
     val eventsAppMetadata: EventsAppMetadata?,
     val enableSensors: Boolean,
-    internal val copilotOptions: CopilotOptions,
+    @OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
```
```suggestion
    @ExperimentalPreviewMapboxNavigationAPI
```

Zayankovsky commented on libnavigation-base/src/main/java/com/mapbox/navigation/base/options/NavigationOptions.kt:
```diff
@@ -42,6 +42,7 @@ const val DEFAULT_NAVIGATOR_PREDICTION_MILLIS = 1000L
  * @param enableSensors enables sensors for current position calculation (optional)
+ * @param copilotOptions defines options for Copilot
  */
 class NavigationOptions private constructor(
```
```suggestion
class NavigationOptions @OptIn(ExperimentalPreviewMapboxNavigationAPI::class) private constructor(
```
Also `toBuilder` and `toString` have to be marked with `@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)`.

Zayankovsky commented on libnavigation-base/src/main/java/com/mapbox/navigation/base/options/NavigationOptions.kt:
```diff
@@ -315,7 +321,8 @@ class NavigationOptions private constructor(
          * Defines configuration for Copilot
          */
-        internal fun copilotOptions(copilotOptions: CopilotOptions): Builder =
+        @OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
```
```suggestion
        @ExperimentalPreviewMapboxNavigationAPI
```

brandanblevins commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilot.kt:
```diff
@@ -0,0 +1,92 @@
+import java.util.concurrent.CopyOnWriteArraySet
+
+/**
+ * Mapbox Copilot is a Navigation SDK component that collects detailed trace files of navigation sessions together with search analytics data.
```
Should it be "Copilot" or "Co-pilot" (or maybe "CoPilot" like Github)?

I see Copilot being used consistently already, so no problems there. Just checking for my own sake.

Guardiola31337 replied:
Yeah Copilot 👍

brandanblevins commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilot.kt:
```diff
@@ -0,0 +1,92 @@
+/**
+ * Mapbox Copilot is a Navigation SDK component that collects detailed trace files of navigation sessions together with search analytics data.
+ *
+ * Mapbox Copilot collects information that helps customers in two different ways:
```
Once "Mapbox Copilot" has been used at the top of the document, you can probably switch to just "Copilot" for the sake of brevity the rest of way.

brandanblevins replied:
```suggestion
 * Mapbox Copilot collects information that helps you in two ways:
```

brandanblevins commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilot.kt:
```diff
@@ -0,0 +1,92 @@
+ * Mapbox Copilot is a Navigation SDK component that collects detailed trace files of navigation sessions together with search analytics data.
+ *
+ * Mapbox Copilot collects information that helps customers in two different ways:
+ * - It makes the feedback provided by drivers more actionable.
```
```suggestion
 * - It makes driver feedback more actionable.
```

brandanblevins commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilot.kt:
```diff
@@ -0,0 +1,92 @@
+ * - It makes the feedback provided by drivers more actionable.
+ * - It provides Mapbox teams with the necessary information to investigate and resolve driver-reported issues and, consequently, helps expedite SDK updates that improve the driver experience for the entire developer community.
+ *
+ * It's a [MapboxNavigationObserver] so it's tied to [MapboxNavigation] lifecycle automatically.
```
```suggestion
 * Copilot is a [MapboxNavigationObserver], so it's tied to the [MapboxNavigation] lifecycle automatically.
```

brandanblevins commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilot.kt:
```diff
@@ -0,0 +1,92 @@
+ * - It provides Mapbox teams with the necessary information to investigate and resolve driver-reported issues and, consequently, helps expedite SDK updates that improve the driver experience for the entire developer community.
+ *
+ * It's a [MapboxNavigationObserver] so it's tied to [MapboxNavigation] lifecycle automatically.
+ * It's highly recommended to keep track of the [DeveloperMetadata.copilotSessionId] (see [DeveloperMetadataObserver]) so that Mapbox teams can easily act on specific end-users feedback as Mapbox teams can easily find the respective traces and troubleshoot issues faster.
```
```suggestion
 * We recommended tracking the [DeveloperMetadata.copilotSessionId] (see [DeveloperMetadataObserver]) so that Mapbox teams can better act on specific end-user feedback. This ID helsp Mapbox teams find the respective traces and troubleshoot issues faster.
```

brandanblevins commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilot.kt:
```diff
@@ -0,0 +1,92 @@
+ * It's highly recommended to keep track of the [DeveloperMetadata.copilotSessionId] (see [DeveloperMetadataObserver]) so that Mapbox teams can easily act on specific end-users feedback as Mapbox teams can easily find the respective traces and troubleshoot issues faster.
+ *
+ * Mapbox Copilot is an opt-in by the customer (see [start] and [stop]) i.e. Navigation SDK customers have the choice to enable it for their drivers.
+ * Depending on the use case, some customers can enable Copilot to all drivers (e.g. during a pilot) or just a subset of them.
```
```suggestion
 * Depending on the use case, you can enable Copilot for either all drivers (for example, during a pilot) or a subset of them.
```

brandanblevins commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilot.kt:
```diff
@@ -0,0 +1,92 @@
+ *
+ * Mapbox Copilot is an opt-in by the customer (see [start] and [stop]) i.e. Navigation SDK customers have the choice to enable it for their drivers.
+ * Depending on the use case, some customers can enable Copilot to all drivers (e.g. during a pilot) or just a subset of them.
+ * **The application developer is responsible for communicating to their drivers when and what data is being collected from their drives.**
```
```suggestion
 * **As the application developer, you are responsible for communicating to drivers about the data that is being collected from their drives, including what kind of data is being collected and when it is collected.**
```

brandanblevins commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilot.kt:
```diff
@@ -0,0 +1,92 @@
+ *
+ * Nav SDK exposes configuration settings (see [NavigationOptions.copilotOptions]) to use Copilot in two ways:
+ * 1) Automatic data collection:
+ * - Copilot will be enabled for all trips performed by a specific driver (default option).
```
```suggestion
 * - Enable Copilot for all trips performed by a specific driver (default option).
```

brandanblevins commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilot.kt:
```diff
@@ -0,0 +1,92 @@
+ * - Copilot will be enabled for all trips performed by a specific driver (default option).
+ * 2) Manual data collection:
+ * - Copilot data will only be sent when an end user submits negative feedback about a specific route to help take action on the issue.
+ * Noting that this is tightly coupled to the Navigation SDK Feedback i.e. this is only effective if the feedback events are pushed through [MapboxNavigation] Feedback APIs (see [MapboxNavigation.postUserFeedback] and [MapboxNavigation.provideFeedbackMetadataWrapper]).
```
```suggestion
 * Data collection for Copilot is tightly coupled to the Navigation SDK Feedback, which means this is only effective if the feedback events are pushed through [MapboxNavigation] Feedback APIs (see [MapboxNavigation.postUserFeedback] and [MapboxNavigation.provideFeedbackMetadataWrapper]).
```

brandanblevins reviewed (changes requested):
See comments and suggestions for MapboxCopilot.kt, and let me know I can assist further. Thanks!

Guardiola31337 commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilot.kt:
```diff
@@ -0,0 +1,92 @@
+ * - Copilot data will only be sent when an end user submits negative feedback about a specific route to help take action on the issue.
+ * Noting that this is tightly coupled to the Navigation SDK Feedback i.e. this is only effective if the feedback events are pushed through [MapboxNavigation] Feedback APIs (see [MapboxNavigation.postUserFeedback] and [MapboxNavigation.provideFeedbackMetadataWrapper]).
+ *
+ * If you would like to provide Search analytics into Copilot, you can send the Search events over to Copilot (see [push]).
```
Yes, that's correct.

Zayankovsky commented on libnavigation-copilot/src/main/java/com/mapbox/navigation/copilot/MapboxCopilotImpl.kt:
```diff
@@ -0,0 +1,391 @@
+     */
+    fun start() {
+        registerUserFeedbackCallback(userFeedbackCallback)
+        ProcessLifecycleOwner.get().lifecycle.addObserver(foregroundBackgroundLifecycleObserver)
```
Should we use `MapboxNavigationApp.lifecycleOwner`? This way we will also support Android Auto.
